File: app/api/routes/documents.py
"""Document routes."""

from typing import Annotated

from uuid import UUID

from fastapi import APIRouter, File, Form, Query, UploadFile, status

from app.api.deps import (
    DocumentServiceDep,
    RequireDocumentMutateDep,
    RequireViewDocumentsDep,
)
from app.infrastructure.db.enums import DocumentStatus
from app.modules.documents.schemas import (
    DocumentDetailResponse,
    DocumentListResponse,
    DocumentUploadResponse,
)

router = APIRouter(prefix="/workspaces", tags=["documents"])


@router.get(
    "/{workspace_id}/documents",
    response_model=DocumentListResponse,
)
async def list_documents(
    workspace_context: RequireViewDocumentsDep,
    document_service: DocumentServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    status: Annotated[DocumentStatus | None, Query()] = None,
) -> DocumentListResponse:
    """List documents in the workspace with optional status filter."""
    return document_service.list_documents(
        context=workspace_context,
        page=page,
        page_size=page_size,
        status=status,
    )


@router.get(
    "/{workspace_id}/documents/{document_id}",
    response_model=DocumentDetailResponse,
)
async def get_document(
    document_id: UUID,
    workspace_context: RequireViewDocumentsDep,
    document_service: DocumentServiceDep,
) -> DocumentDetailResponse:
    """Return document details including ingestion status and chunk count."""
    return document_service.get_document_detail(
        context=workspace_context,
        document_id=document_id,
    )


@router.post(
    "/{workspace_id}/documents",
    response_model=DocumentUploadResponse,
    status_code=status.HTTP_202_ACCEPTED,
)
async def upload_document(
    workspace_context: RequireDocumentMutateDep,
    document_service: DocumentServiceDep,
    file: Annotated[UploadFile, File()],
    title: Annotated[str | None, Form()] = None,
    source_type: Annotated[str | None, Form()] = None,
) -> DocumentUploadResponse:
    """Upload a supported document to the workspace."""
    content = await file.read()
    return document_service.upload(
        context=workspace_context,
        file_content=content,
        filename=file.filename or "untitled",
        title=title,
        source_type=source_type,
    )

File: web/src/lib/api/documents.ts
import { apiFetch } from "@/lib/api/client";
import type {
  DocumentListResponse,
  DocumentUploadResponse,
} from "@/lib/api/types";

export type ListDocumentsParams = {
  page?: number;
  pageSize?: number;
  status?: string;
};

/** GET /workspaces/{id}/documents */
export function listDocuments(
  workspaceId: string,
  params: ListDocumentsParams = {},
): Promise<DocumentListResponse> {
  const query = new URLSearchParams();
  if (params.page != null) {
    query.set("page", String(params.page));
  }
  if (params.pageSize != null) {
    query.set("page_size", String(params.pageSize));
  }
  if (params.status) {
    query.set("status", params.status);
  }
  const qs = query.toString();
  const path = `/workspaces/${workspaceId}/documents${qs ? `?${qs}` : ""}`;
  return apiFetch<DocumentListResponse>(path);
}

/**
 * POST /workspaces/{id}/documents — multipart upload.
 * Field name `file` matches the FastAPI UploadFile parameter.
 */
export function uploadDocument(
  workspaceId: string,
  file: File,
  options: { title?: string; sourceType?: string } = {},
): Promise<DocumentUploadResponse> {
  const body = new FormData();
  body.append("file", file);
  if (options.title) {
    body.append("title", options.title);
  }
  if (options.sourceType) {
    body.append("source_type", options.sourceType);
  }
  return apiFetch<DocumentUploadResponse>(
    `/workspaces/${workspaceId}/documents`,
    {
      method: "POST",
      body,
    },
  );
}
